File: Software/src/battery/SONO-BATTERY.cpp
#include "SONO-BATTERY.h"
#include "../communication/can/comm_can.h"
#include "../datalayer/datalayer.h"
#include "../devboard/utils/events.h"
#include "../include.h"

void SonoBattery::
    update_values() {  //This function maps all the values fetched via CAN to the correct parameters used for modbus

  datalayer.battery.status.real_soc = (realSOC * 100);  //increase SOC range from 0-100 -> 100.00

  datalayer.battery.status.soh_pptt = (batterySOH * 100);  //Increase decimals from 100% -> 100.00%

  datalayer.battery.status.voltage_dV = batteryVoltage;

  datalayer.battery.status.current_dA = batteryAmps;

  datalayer.battery.info.total_capacity_Wh = 54000;

  datalayer.battery.status.remaining_capacity_Wh = static_cast<uint32_t>(
      (static_cast<double>(datalayer.battery.status.real_soc) / 10000) * datalayer.battery.info.total_capacity_Wh);

  datalayer.battery.status.max_discharge_power_W = allowedDischargePower * 100;

  datalayer.battery.status.max_charge_power_W = allowedChargePower * 100;

  datalayer.battery.status.cell_max_voltage_mV = CellVoltMax_mV;

  datalayer.battery.status.cell_min_voltage_mV = CellVoltMin_mV;

  datalayer.battery.status.temperature_min_dC = temperatureMin;

  datalayer.battery.status.temperature_max_dC = temperatureMax;
}

void SonoBattery::handle_incoming_can_frame(CAN_frame rx_frame) {
  switch (rx_frame.ID) {
    case 0x100:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x101:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x102:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      functionalsafetybitmask = rx_frame.data.u8[0];  //If any bits are high here, battery has a HSD fault active.
      break;
    case 0x200:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x220:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      allowedChargePower = (rx_frame.data.u8[0] << 8) + rx_frame.data.u8[1];
      allowedDischargePower = (rx_frame.data.u8[2] << 8) + rx_frame.data.u8[3];
      break;
    case 0x221:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x300:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      batteryVoltage = (rx_frame.data.u8[4] << 8) + rx_frame.data.u8[5];
      break;
    case 0x301:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      CellVoltMax_mV = (rx_frame.data.u8[1] << 8) + rx_frame.data.u8[2];
      CellVoltMin_mV = (rx_frame.data.u8[4] << 8) + rx_frame.data.u8[5];
      break;
    case 0x310:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x311:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      batteryAmps = ((rx_frame.data.u8[4] << 8) + rx_frame.data.u8[5]) - 1000;
      break;
    case 0x320:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      temperatureMax = ((rx_frame.data.u8[1] << 8) + rx_frame.data.u8[2]) - 400;
      temperatureMin = ((rx_frame.data.u8[4] << 8) + rx_frame.data.u8[5]) - 400;
      break;
    case 0x321:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      batterySOH = rx_frame.data.u8[4];
      break;
    case 0x330:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x331:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x601:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      realSOC = rx_frame.data.u8[0];
      break;
    case 0x610:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x611:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x613:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x614:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    case 0x615:
      datalayer.battery.status.CAN_battery_still_alive = CAN_STILL_ALIVE;
      break;
    default:
      break;
  }
}
void SonoBattery::transmit_can(unsigned long currentMillis) {
  // Send 100ms CAN Message
  if (currentMillis - previousMillis100 >= INTERVAL_100_MS) {
    previousMillis100 = currentMillis;
    //VCU Command message
    SONO_400.data.u8[0] = 0x15;  //Charging enabled bit01, dischargign enabled bit23, dc charging bit45

    if (datalayer.battery.status.bms_status == FAULT) {
      SONO_400.data.u8[0] = 0x14;  //Charging DISABLED
    }
    transmit_can_frame(&SONO_400, can_config.battery);
  }
  // Send 1000ms CAN Message
  if (currentMillis - previousMillis1000 >= INTERVAL_1_S) {
    previousMillis1000 = currentMillis;

    //Time and date
    //Let's see if the battery is happy with just getting seconds incrementing
    SONO_401.data.u8[0] = 2025;     //Year
    SONO_401.data.u8[1] = 1;        //Month
    SONO_401.data.u8[2] = 1;        //Day
    SONO_401.data.u8[3] = 12;       //Hour
    SONO_401.data.u8[4] = 15;       //Minute
    SONO_401.data.u8[5] = seconds;  //Second
    seconds = (seconds + 1) % 61;
    transmit_can_frame(&SONO_401, can_config.battery);
  }
}

void SonoBattery::setup(void) {  // Performs one time setup at startup
  strncpy(datalayer.system.info.battery_protocol, Name, 63);
  datalayer.system.info.battery_protocol[63] = '\0';
  datalayer.battery.info.number_of_cells = 96;
  datalayer.system.status.battery_allows_contactor_closing = true;
  datalayer.battery.info.max_design_voltage_dV = MAX_PACK_VOLTAGE_DV;
  datalayer.battery.info.min_design_voltage_dV = MIN_PACK_VOLTAGE_DV;
  datalayer.battery.info.max_cell_voltage_mV = MAX_CELL_VOLTAGE_MV;
  datalayer.battery.info.min_cell_voltage_mV = MIN_CELL_VOLTAGE_MV;
  datalayer.battery.info.max_cell_voltage_deviation_mV = MAX_CELL_DEVIATION_MV;
  datalayer.battery.info.chemistry = battery_chemistry_enum::LFP;
}

File: frontend/src/ext/datalayer/SonoBattery.ts
// SonoBattery: 200 bytes; base classes: CanBattery@0
export const SONO_BATTERY_FIELDS: ([string, string] | [string, string, number])[] = [
  ['___vptr$Battery', '__vtbl_ptr_type*'],
  ['__defaultRenderer', 'BatteryDefaultRenderer'],
  ['', ' ', 4],
  ['___vptr$Transmitter', '__vtbl_ptr_type*'],
  ['___vptr$CanReceiver', '__vtbl_ptr_type*'],
  ['_can_interface', 'CAN_Interface'],
  ['_initial_speed', 'CAN_Speed'],
  ['previousMillis100', 'u32'],
  ['previousMillis1000', 'u32'],
  ['seconds', 'u8'],
  ['functionalsafetybitmask', 'u8'],
  ['batteryVoltage', 'u16'],
  ['allowedDischargePower', 'u16'],
  ['allowedChargePower', 'u16'],
  ['CellVoltMax_mV', 'u16'],
  ['CellVoltMin_mV', 'u16'],
  ['batteryAmps', 'i16'],
  ['temperatureMin', 'i16'],
  ['temperatureMax', 'i16'],
  ['batterySOH', 'u8'],
  ['realSOC', 'u8'],
  ['', ' ', 4],
  ['SONO_400', 'CAN_frame'],
  ['SONO_401', 'CAN_frame'],
];

export const ___vptr$Battery: [number, string] = [0, '__vtbl_ptr_type*'];
export const __defaultRenderer: [number, string] = [4, 'BatteryDefaultRenderer'];
export const ___vptr$Transmitter: [number, string] = [8, '__vtbl_ptr_type*'];
export const ___vptr$CanReceiver: [number, string] = [12, '__vtbl_ptr_type*'];
export const _can_interface: [number, string] = [16, 'CAN_Interface'];
export const _initial_speed: [number, string] = [20, 'CAN_Speed'];
export const previousMillis100: [number, string] = [24, 'u32'];
export const previousMillis1000: [number, string] = [28, 'u32'];
export const seconds: [number, string] = [32, 'u8'];
export const functionalsafetybitmask: [number, string] = [33, 'u8'];
export const batteryVoltage: [number, string] = [34, 'u16'];
export const allowedDischargePower: [number, string] = [36, 'u16'];
export const allowedChargePower: [number, string] = [38, 'u16'];
export const CellVoltMax_mV: [number, string] = [40, 'u16'];
export const CellVoltMin_mV: [number, string] = [42, 'u16'];
export const batteryAmps: [number, string] = [44, 'i16'];
export const temperatureMin: [number, string] = [46, 'i16'];
export const temperatureMax: [number, string] = [48, 'i16'];
export const batterySOH: [number, string] = [50, 'u8'];
export const realSOC: [number, string] = [51, 'u8'];
export const SONO_400: [number, string] = [56, 'CAN_frame'];
export const SONO_401: [number, string] = [128, 'CAN_frame'];
